packages/tpetra/core/src/Tpetra_Apply_Helpers.hpp

```diff
@@ -0,0 +1,192 @@
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_APPLY_HELPERS_HPP
+#define TPETRA_APPLY_HELPERS_HPP
+#include <type_traits>
+#include "Teuchos_ScalarTraits.hpp"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+
+namespace Tpetra {
+  //! @name Methods performing multiple matrix-vector products at once
+  //@{
+  /// \brief Does multiply matrix apply() calls with a single X vector
+  ///
+  /// Computes Y[i] = Matrices[i] * X, for i=1,...,N
+  ///
+  /// This routine only communicates the interprocessor portions of X once,
+  /// if such a thing is possible (aka the ColMap's of the matrices match). 
+  ///
+  /// If the Y's are replicated this will not use the reduced communication path.
+  ///
+  /// If X is aliased to any Y but the last one, this routine will throw
+  ///
+  /// \param Matrices [in] - [std::vector|Teuchos::Array|Teuchos::ArrayRCP] of Tpetra::CrsMatrix objects.
+  ///                        These matrices can different numbers of rows. 
+  /// \param X [in]        - Tpetra::MultiVector or Tpetra::Vector object.
+  /// \param Y [out]       - [std::vector|Teuchos::Array|Teuchos::ArrayRCP] of Tpetra::MultiVector or Tpetra::Vector objects.
+  ///                        These must have the same number of vectors as X.
+  template <class MatrixArray, class MultiVectorArray> 
+  void batchedApply(const MatrixArray &Matrices, 
+                    const typename std::remove_pointer<typename MultiVectorArray::value_type>::type &X,
+                    MultiVectorArray &Y,
+                    typename std::remove_pointer<typename MatrixArray::value_type>::type::scalar_type alpha = Teuchos::ScalarTraits< typename std::remove_pointer<typename MatrixArray::value_type>::type::scalar_type>::one(),
+                    typename std::remove_pointer<typename MatrixArray::value_type>::type::scalar_type beta  = Teuchos::ScalarTraits< typename std::remove_pointer<typename MatrixArray::value_type>::type::scalar_type>::zero() ) {
+    Tpetra::Details::ProfilingRegion regionImport ("Tpetra::batchedApply");
```

Comment:
`regionImport` needs renaming to avoid clash with the other timer below.

Reply:
Good catch!